src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.cs

```diff
@@ -66,92 +65,170 @@ public async Task FindReferencesAsync(ISymbol symbol, CancellationToken cancella
-                var projectToDocumentMap = await CreateProjectToDocumentMapAsync(projectMap, cancellationToken).ConfigureAwait(false);
-                ValidateProjectToDocumentMap(projectToDocumentMap);
+                    var currentProject = _solution.GetRequiredProject(projectId);
 
-                await ProcessAsync(projectToDocumentMap, cancellationToken).ConfigureAwait(false);
+                    // As we walk each project, attempt to grow the search set appropriately up and down the inheritance
+                    // hierarchy and grab a copy of the symbols to be processed.  Note: this has to happen serially
+                    // which is why we do it in this loop and not inside the concurrent project processing that happens
+                    // below.
+                    await symbolSet.InheritanceCascadeAsync(currentProject, cancellationToken).ConfigureAwait(false);
+                    allSymbols = symbolSet.GetAllSymbols();
+
+                    // Report any new symbols we've cascaded to to our caller.
+                    await ReportGroupsAsync(allSymbols, cancellationToken).ConfigureAwait(false);
+
+                    tasks.Add(CreateWorkAsync(() => ProcessProjectAsync(currentProject, allSymbols, cancellationToken), cancellationToken));
+                }
+
+                // Now, wait for all projects to complete.
+                await Task.WhenAll(tasks).ConfigureAwait(false);
             }
             finally
             {
                 await _progress.OnCompletedAsync(cancellationToken).ConfigureAwait(false);
             }
         }
 
-        private async Task ProcessAsync(ProjectToDocumentMap projectToDocumentMap, CancellationToken cancellationToken)
+        public Task CreateWorkAsync(Func<Task> createWorkAsync, CancellationToken cancellationToken)
+            => Task.Factory.StartNew(createWorkAsync, cancellationToken, TaskCreationOptions.None, _scheduler).Unwrap();
+
+        /// <summary>
+        /// Notify the caller of the engine about the definitions we've found that we're looking for.  We'll only notify
+        /// them once per symbol group, but we may have to notify about new symbols each time we expand our symbol set
+        /// when we walk into a new project.
+        /// </summary>
+        private async Task ReportGroupsAsync(ImmutableArray<ISymbol> symbols, CancellationToken cancellationToken)
         {
-            using (Logger.LogBlock(FunctionId.FindReference_ProcessAsync, cancellationToken))
+            foreach (var symbol in symbols)
             {
-                // quick exit
-                if (projectToDocumentMap.Count == 0)
+                if (!_symbolToGroup.ContainsKey(symbol))
                 {
-                    return;
-                }
+                    var linkedSymbols = await SymbolFinder.FindLinkedSymbolsAsync(symbol, _solution, cancellationToken).ConfigureAwait(false);
```

Comment:
Shouldn't the symbol set already include all the linked symbols as they are added during initial creation and cascading via `AddCascadedAndLinkedSymbolsToAsync`

Reply:
Yes. The symbol *set* has all the symbols. But what we're doing here is reporting the symbol groups (the clique of linked symbols). That said, what I can likely do is make this entire system operate on symbol groups instead. That might clean things up.  I'll see what I can do.

Reply:
it did not clean things up :)